src/inmanta/deploy/state.py

```diff
@@ -36,15 +38,45 @@ def requires_view(self) -> Mapping[ResourceIdStr, Set[ResourceIdStr]]:
     def provides_view(self) -> Mapping[ResourceIdStr, Set[ResourceIdStr]]:
         return self.reverse_mapping()
 
+    def get_all_provides_transitively(self, resource: ResourceIdStr | set[ResourceIdStr]) -> list[ResourceIdStr]:
+        """
+        This method returns all the provides (transitively) of the given resource.
+
+        :param resource: The resource or set of resources for which the provides have to be resolved transitively.
+        """
+        input_set = {resource} if isinstance(resource, str) else set(resource)
+        work: SimpleQueue[ResourceIdStr] = SimpleQueue()
+        for elem in input_set:
+            work.put_nowait(elem)
+        provides_mapping = self.provides_view()
+        # Use a dict here to not lost the order of the elements.
+        result: dict[ResourceIdStr, None] = {}
+        while not work.empty():
+            current_resource = work.get()
+            result[current_resource] = None
+            provides = provides_mapping.get(current_resource, set())
+            for elem in provides:
+                work.put_nowait(elem)
```

Comment:
Only if it's not yet in `result`, right? We might even need a set to keep track of the union of result and work if we don't want to queue the same element a bunch of times.